module-6/src/repositories/cart.repository.ts

```diff
@@ -1,58 +1,113 @@
+import { ORDER_STATUS } from "../entities/Order";
+
+import { DI } from "../index";
 
 const getUserCartByUserId = async (
   userId: string
 ): Promise<CartEntity | null> => {
-  return await Cart.findOne({ userId });
+  const cart = await DI.cartRepository.findOne(
+    { user: { id: userId } },
+    { populate: ["items", "items.product"] }
+  );
+  if (cart) {
+    return {
+      id: cart.id,
+      userId: cart.user.id,
+      items: cart.items.toArray().map(({ product, id, count }) => {
+        return { id, product, count };
+      }),
+    };
+  }
+  return null;
 };
 
 const createUserCart = async (userId: string): Promise<CartEntity> => {
-  return await Cart.create({ userId, isDeleted: false, items: [] });
+  const cart = DI.cartRepository.create({
+    isDeleted: false,
+    items: [],
+    user: { id: userId },
+  });
+
+  DI.em.persistAndFlush(cart);
+
+  return {
+    id: cart.id,
+    items: [],
+    userId: cart.user.id,
+  };
 };
 
-const updateUserCart = async (userId: string, cartItem: CartItemEntity) => {
-  const cart = await Cart.findOne({ userId });
+const updateUserCart = async (
+  userId: string,
+  { count, product }: CartItemEntity
+) => {
+  const cart = await DI.cartRepository.findOne(
+    { user: { id: userId } },
+    { populate: ["items", "items.product"] }
+  );
 
   if (cart) {
-    cart.items = [
-      ...cart.items.filter((item) => item.product.id !== cartItem.product.id),
-      cartItem,
-    ];
+    const existingCartItem = cart.items.find(
+      (item) => item.product.id === product.id
+    );
 
-    await cart.save();
-  }
+    if (existingCartItem) {
+      existingCartItem.count = count;
+    } else {
+      const cartItem = await DI.cartItemRepository.create({
+        count,
+        product,
+        cart,
+      });
+      cart.items.add(cartItem);
+    }
+
+    await DI.em.flush();
 
-  return cart;
+    return {
+      id: cart.id,
+      userId: cart.user.id,
+      items: cart.items.toArray().map(({ product, id, count }) => {
+        return { id, product, count };
+      }),
+    };
+  }
 };
 
 const deleteUserCart = async (userId: string) => {
-  await Cart.updateOne({ userId }, { items: [] });
+  const cart = await DI.cartRepository.findOne({ user: { id: userId } });
+  const cartItems = await DI.cartItemRepository.find({ cart: cart?.id });
+
+  await DI.em.removeAndFlush(cartItems);
 };
 
-const createOrder = async (
-  userId: string,
-  cartId: string,
-  items: CartItemEntity[],
-  total: number
-) => {
-  return await Order.create({
-    userId,
-    cartId,
-    items,
-    total,
+const createOrder = async (userId: string, total: number) => {
+  const userCart = await DI.cartRepository.findOne(
+    { user: { id: userId } },
+    { populate: ["items", "items.product"] }
+  );
+
+  const user = (await DI.userRepository.findOne({ id: userId })) as User;
+
+  const order = DI.orderRepository.create({
+    user,
+    items: userCart?.items,
     payment: { type: "test" },
```

Comment:
Ensure all fields, especially like payment and delivery, are validated before creating an order.